modules/bullet/space_bullet.cpp

```diff
@@ -349,14 +358,67 @@ SpaceBullet::~SpaceBullet() {
 	destroy_world();
 }
 
+void SpaceBullet::add_to_pre_flush_queue(CollisionObjectBullet *p_co) {
+	if (p_co->is_in_flush_queue == false) {
+		p_co->is_in_flush_queue = true;
+		queue_pre_flush.push_back(p_co);
+	}
+}
+
+void SpaceBullet::add_to_flush_queue(CollisionObjectBullet *p_co) {
+	if (p_co->is_in_flush_queue == false) {
+		p_co->is_in_flush_queue = true;
+		queue_flush.push_back(p_co);
+	}
+}
+
+void SpaceBullet::add_to_dirty_queue(CollisionObjectBullet *p_co) {
+	if (p_co->is_in_dirty_queue == false) {
+		p_co->is_in_dirty_queue = true;
+		queue_dirty.push_back(p_co);
+	}
+}
+
+void SpaceBullet::remove_from_any_queue(CollisionObjectBullet *p_co) {
+	if (p_co->is_in_flush_queue) {
+		p_co->is_in_flush_queue = false;
+		queue_pre_flush.erase(p_co);
+		queue_flush.erase(p_co);
+	}
+	if (p_co->is_in_dirty_queue) {
+		p_co->is_in_dirty_queue = false;
+		queue_dirty.erase(p_co);
+	}
+}
+
 void SpaceBullet::flush_queries() {
-	const btCollisionObjectArray &colObjArray = dynamicsWorld->getCollisionObjectArray();
-	for (int i = colObjArray.size() - 1; 0 <= i; --i) {
-		static_cast<CollisionObjectBullet *>(colObjArray[i]->getUserPointer())->dispatch_callbacks();
+	for (uint32_t i = 0; i < queue_pre_flush.size(); i += 1) {
+		queue_pre_flush[i]->dispatch_callbacks();
+		queue_pre_flush[i]->is_in_flush_queue = false;
+	}
+	for (uint32_t i = 0; i < queue_flush.size(); i += 1) {
+		queue_flush[i]->dispatch_callbacks();
+		queue_flush[i]->is_in_flush_queue = false;
 	}
+	queue_pre_flush.clear();
+	queue_flush.clear();
```

Comment:
Some comments here to explain what the logic is behind having two different queues could make it clearer for future contributors.

Reply:
Done.

Reply:
yes please more comments for future contributors

Reply:
Already done. You can find it above: https://github.com/godotengine/godot/pull/42643/files#diff-28f7bca21ebbe699cd16802c05d43f4e085331c7d0427c38da0ab3e6d14dc696R395-R399

Reply:
👍 awesome thx!